### PR TITLE
Refactor missing discrete dists

### PR DESCRIPTION
**Pull request by ricardoV94**
This PR refactors all missing univariate discrete distributions.

1. I decided to refactor the `Constant` distribution after some debate on Slack as to whether this should be deprecated or not. Its direct use might be pretty problematic without a custom step algorithm, but could still be useful indirectly, such as in mixture distributions. I made the dtype of the RV be `floatX` to accept non-integer outputs (even though this distribution should be treated as a pmf, see https://github.com/pymc-devs/pymc3/issues/4511 for why this might matter)

2. I created separate RandomOps for the ZeroInflated* distributions for now. Similarly to #4683, this should only be done temporarily until we have Mixture Distributions working on V4 (it seems such approach was already attempted in the past #2246, #1459)
 
***

Depending on what your PR does, here are a few things you might want to address in the description:
+ [x] what are the (breaking) changes that this PR makes?
+ [x] important background, or details about the implementation
+ [x] are the changes—especially new features—covered by tests and docstrings?
+ [x] [linting/style checks have been run](https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style)
+ [x] [consider adding/updating relevant example notebooks](https://github.com/pymc-devs/pymc-examples)
+ [x] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

### CODE REVIEW

File: pymc3/distributions/discrete.py [hunk 1]
```diff
@@ -1177,6 +1161,22 @@ def logp(value, p):
...
+    name = "constant"
+    ndim_supp = 0
+    ndims_params = [0]
+    dtype = "floatX"  # Should be treated as a discrete variable!
+    _print_name = ("Constant", "\\operatorname{Constant}")
+
+    @classmethod
+    def rng_fn(cls, rng, c, size=None):
+        ones = np.ones_like(c)
+        return rng.randint(ones, ones + 1, size=size) * c
```
Comment (ricardoV94) [inline pymc3/distributions/discrete.py]: There must be a less hacking way to get a filled array given `c` and `size`, that works as other random generators would. `np.full(size, c)` doesn't seem to cut it, because it works with shape and not with size.

Comment (twiecki) [reply pymc3/distributions/discrete.py]: Isn't shape and size the same thing here?

**Review by twiecki — DISMISSED** [review]
This looks great!

**Comment by twiecki** [comment]
Thanks for taking this on @ricardoV94. Separately, when I tried V4 it seems like other continuous distributions are also missing, like `StudentT`?

File: pymc3/distributions/discrete.py (hunk 1 above)
Comment (ricardoV94) [reply pymc3/distributions/discrete.py]: No because `c` could be a vector and `size` could be None for instance

**Comment by ricardoV94** [comment]
Yes they are missing.

Only distributions that have a `rv_op` at the top have been refactored. StudentT hasn't yet.

I won't have more time this week, so if someone wants to open a PR for the missing continuous distributions it would be great. The logic should be pretty similar to what is done in this PR with the custom RandomOps

File: pymc3/distributions/discrete.py [hunk 2]
```diff
@@ -1335,18 +1302,22 @@ def logp(self, value):
         -------
         TensorVariable
         """
-        psi = self.psi
-        theta = self.theta
 
         logp_val = at.switch(
             at.gt(value, 0),
-            at.log(psi) + self.pois.logp(value),
+            at.log(psi) + Poisson.logp(value, theta),
```
Comment (brandonwillard) [inline pymc3/distributions/discrete.py]: Those `logp` functions inside the `Distribution` classes are just a convenience for declaring dispatch functions, so we don't want to rely on them in any way.  Instead, call the underlying `_logp` directly.

File: pymc3/distributions/discrete.py [hunk 3]
```diff
@@ -1361,16 +1332,31 @@ def logcdf(self, value):
         -------
         TensorVariable
         """
-        psi = self.psi
 
         return bound(
-            logaddexp(at.log1p(-psi), at.log(psi) + self.pois.logcdf(value)),
+            logaddexp(at.log1p(-psi), at.log(psi) + Poisson.logcdf(value, theta)),
```
Comment (brandonwillard) [inline pymc3/distributions/discrete.py]: See my comment above about using `_logp` instead.

File: pymc3/distributions/discrete.py [hunk 4]
```diff
@@ -1467,19 +1432,24 @@ def logp(self, value):
...
         TensorVariable
         """
-        psi = self.psi
-        p = self.p
-        n = self.n
 
         logp_val = at.switch(
             at.gt(value, 0),
-            at.log(psi) + self.bin.logp(value),
+            at.log(psi) + Binomial.logp(value, n, p),
```
Comment (brandonwillard) [inline pymc3/distributions/discrete.py]: See my comment above about using `_logp` instead.

File: pymc3/distributions/discrete.py [hunk 5]
```diff
@@ -1493,22 +1463,39 @@ def logcdf(self, value):
...
         if np.ndim(value):
             raise TypeError(
                 f"ZeroInflatedBinomial.logcdf expects a scalar value but received a {np.ndim(value)}-dimensional object."
             )
 
-        psi = self.psi
-
         return bound(
-            logaddexp(at.log1p(-psi), at.log(psi) + self.bin.logcdf(value)),
+            logaddexp(at.log1p(-psi), at.log(psi) + Binomial.logcdf(value, n, p)),
```
Comment (brandonwillard) [inline pymc3/distributions/discrete.py]: See my comment above about using `_logp` instead.

File: pymc3/distributions/discrete.py [hunk 6]
```diff
@@ -1635,20 +1589,22 @@ def logp(self, value):
...
-        mu = self.mu
-        psi = self.psi
 
-        logp_other = at.log(psi) + self.nb.logp(value)
-        logp_0 = logaddexp(
-            at.log1p(-psi), at.log(psi) + alpha * (at.log(alpha) - at.log(alpha + mu))
+        return bound(
+            at.switch(
+                at.gt(value, 0),
+                at.log(psi) + NegativeBinomial.logp(value, n, p),
```
Comment (brandonwillard) [inline pymc3/distributions/discrete.py]: See my comment above about using `_logp` instead.

File: pymc3/distributions/discrete.py [hunk 7]
```diff
@@ -1667,13 +1623,14 @@ def logcdf(self, value):
             raise TypeError(
                 f"ZeroInflatedNegativeBinomial.logcdf expects a scalar value but received a {np.ndim(value)}-dimensional object."
             )
-        psi = self.psi
 
         return bound(
-            logaddexp(at.log1p(-psi), at.log(psi) + self.nb.logcdf(value)),
+            logaddexp(at.log1p(-psi), at.log(psi) + NegativeBinomial.logcdf(value, n, p)),
```
Comment (brandonwillard) [inline pymc3/distributions/discrete.py]: See my comment above about using `_logp` instead.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Looks good, but we really shouldn't rely on `Distribution.logp` functions, since those aren't officially part of the `Distribution` API&mdash;they're just a means of conveniently specifying dispatches.

File: pymc3/distributions/discrete.py (hunk 1 above)
Comment (ricardoV94) [reply pymc3/distributions/discrete.py]: Okay, actually I think the only problem was `size=None`. I added a check for that condition...

**Comment by ricardoV94** [comment]
> Looks good, but we really shouldn't rely on `Distribution.logp` functions, since those aren't officially part of the `Distribution` API—they're just a means of conveniently specifying dispatches.

@brandonwillard I am not sure what you have in mind. This looks wrong, right?

```python
# at.log(psi) + Poisson.logp(value, theta)
at.log(psi) + _logp(poisson, value, {value: value}, theta)
```

I also tried this, but it fails following a `change_rv_size` operation:

```python
at.log(psi) + logpt(Poisson.dist(theta), value)
```

**Comment by brandonwillard** [comment]
> @brandonwillard I am not sure what you have in mind. This looks wrong, right?
> 
> ```python
> # at.log(psi) + Poisson.logp(value, theta),
> at.log(psi) + _logp(poisson, value, {value: value}, theta),
> ```

No, that looks like the correct approach.

**Comment by ricardoV94** [comment]
Could you give some intuition why this is better than the original approach?

**Comment by brandonwillard** [comment]
> Could you give some intuition why this is better than the original approach?

As I stated in my original review comments, these `Distribution.logp` functions are not a part of the `Distribution` API and they are neither designed nor intended to be used in this way.  Their availability is questionable and subject to arbitrary change, so we don't want to use them outside of their intended purpose.

More importantly, we do *not* want to rely on the `Distribution` types in general, and especially not for log-likelihoods; they're currently just a legacy transition artifact.  This is clearly demonstrated by the lack of `Distribution` in the log-likelihood code itself.  

**Simply put, this approach unnecessarily adds `Distribution`s into the log-likelihood framework.**

If this sort of use case needs to be simplified, it can be done in better ways that are based only on the underlying dispatch.

**Review by brandonwillard — APPROVED** [review]

**Comment by twiecki** [comment]
🥳
